orttraining/orttraining/python/training/ortmodule/_inference_manager.py

```diff
@@ -61,40 +66,59 @@ def forward(self, *inputs, **kwargs):
         Finally, we instantiate the ONNX Runtime InferenceSession through the InferenceAgent.
         '''
 
-        # Exporting module to ONNX for the first time
-        build_graph = self._export_model(*inputs, **kwargs)
-        if build_graph:
-            # If model was exported, then initialize the graph builder
-            self._initialize_graph_builder(training=False)
-
-        # Build the inference graph
-        if build_graph:
-            self._build_graph()
-
-        module_device = _utils.get_device_from_module(self._original_module)
-        # The inference session should be created every time
-        # the graph was built or if the device changed between calls to forward
-        create_execution_session = build_graph or self._device != module_device
-        if self._device != module_device:
-            self._device = module_device
-        if create_execution_session:
-            # Create execution session creates the inference_session
-            self._create_execution_agent()
-
-        user_outputs, _ = InferenceManager.execution_session_run_forward(self._execution_agent,
-                                                                         self._onnx_models.optimized_model,
-                                                                         self._device,
-                                                                         *_io._combine_input_buffers_initializers(
-                                                                             self._graph_initializers,
-                                                                             self._graph_info.user_input_names,
-                                                                             self._input_info,
-                                                                             self._flattened_module.named_buffers(),
-                                                                             inputs,
-                                                                             kwargs,
-                                                                             self._device))
-
-        return _io.unflatten_user_output(self._module_output_schema,
-                                         user_outputs)
+        # Fallback to PyTorch due to failures *external* to forward(),
+        #  typically from initialization
```

Comment:
What type of initialization error would result in fallback here?

Reply:
This check catches is intended to catch initialization failures, such as invalid models (e.g. torch.nn.DataParallel), which is not yet fully implemented/tested. It can also detect multiple devices scenarios, although the device check was (temporarily) moved to the forward call for now. I may move this check to __init__ in the next commit